File: server/src/album/album.service.ts
import { Injectable } from '@nestjs/common';
import { InjectModel } from '@nestjs/mongoose';
import { Model, ObjectId } from 'mongoose';
import { FileService, FileType } from '../file/file.service';
import { CreateAlbumDto } from './dto/create-album.dto';
import { Album, AlbumDocument } from './schemas/album.schema';
import { Song, SongDocument } from '../song/schemas/song.schema';
import { User, UserDocument } from '../users/schemas/user.schema';

@Injectable()
export class AlbumService {
  constructor(
    @InjectModel(Song.name) private songModel: Model<SongDocument>,
    @InjectModel(Album.name) private albumModel: Model<AlbumDocument>,
    @InjectModel(User.name) private userModel: Model<UserDocument>,
    private fileService: FileService,
  ) {}

  async create(dto: CreateAlbumDto, user, picture): Promise<Album> {
    const picturePath = this.fileService.createFile(FileType.IMAGE, picture);

    const updateUser = await this.userModel.findById(user);

    const [album] = await Promise.all([
      this.albumModel.create({
        ...dto,
        user: user,
        picture: picturePath,
      }),
    ]);

    updateUser.albums.push(album._id);
    await updateUser.save();
    return album;
  }

  async getAll(count = 10, offset = 0): Promise<Album[]> {
    const [albums] = await Promise.all([
      this.albumModel.find().skip(offset).limit(count),
    ]);
    return albums;
  }

  async getOne(id: ObjectId): Promise<Album> {
    const [album] = await Promise.all([
      this.albumModel.findById(id).populate('songs'),
    ]);
    return album;
  }

  async delete(id: ObjectId): Promise<ObjectId> {
    const album = await this.albumModel.findByIdAndDelete(id);

    for (const songId of album.songs) {
        const song = await this.songModel.findById(songId);
        song.album = null;
        await song.save();
    }

    this.fileService.removeFile(album.picture);
    return album._id;
  }

  async addSong(id: ObjectId, songId: ObjectId): Promise<Album> {
    const album = await this.albumModel.findById(id);
    const song = await this.songModel.findById(songId);
    album.songs.push(song._id);

    song.album = album;
    await album.save();
    await song.save();
    return album;
  }

  async search(query: string): Promise<Album[]> {
    const [albums] = await Promise.all([
      this.albumModel.find({
        $or: [
          { name: { $regex: new RegExp(query, 'i') } },
          { author: { $regex: new RegExp(query, 'i') } },
        ],
      }),
    ]);
    return albums;
  }

  async getCount() {
    const albumCount = await this.albumModel.countDocuments();
    return albumCount;
  }
}

File: server-java/src/main/java/com/muz/service/AlbumService.java
package com.muz.service;

import com.muz.document.AlbumDocument;
import com.muz.document.SongDocument;
import com.muz.document.UserDocument;
import com.muz.dto.CreateAlbumDto;
import com.muz.exception.ApiException;
import com.muz.repository.AlbumRepository;
import com.muz.repository.SongRepository;
import com.muz.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.PageRequest;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.web.multipart.MultipartFile;

import java.util.List;
import java.util.Map;

@Service
@RequiredArgsConstructor
public class AlbumService {

    private final AlbumRepository albumRepository;
    private final SongRepository songRepository;
    private final UserRepository userRepository;
    private final FileStorageService fileStorage;

    public AlbumDocument create(CreateAlbumDto dto, String userId, MultipartFile picture) {
        String picturePath = fileStorage.store("image", picture);

        AlbumDocument album = new AlbumDocument();
        album.setName(dto.getName());
        album.setAuthor(dto.getAuthor());
        album.setPicture(picturePath);
        album.setUser(userId);
        AlbumDocument saved = albumRepository.save(album);

        UserDocument user = userRepository.findById(userId)
                .orElseThrow(() -> new ApiException(HttpStatus.NOT_FOUND, "User not found"));
        user.getAlbums().add(saved.getId());
        userRepository.save(user);

        return saved;
    }

    public List<AlbumDocument> getAll(int count, int offset) {
        return albumRepository.findAll(PageRequest.of(offset / count, count)).getContent();
    }

    public Map<String, Object> getOne(String id) {
        AlbumDocument album = albumRepository.findById(id)
                .orElseThrow(() -> new ApiException(HttpStatus.NOT_FOUND, "Album not found"));
        List<SongDocument> songs = songRepository.findAllById(album.getSongs());
        return Map.of("album", album, "songs", songs);
    }

    public String delete(String id) {
        AlbumDocument album = albumRepository.findById(id)
                .orElseThrow(() -> new ApiException(HttpStatus.NOT_FOUND, "Album not found"));

        for (String songId : album.getSongs()) {
            songRepository.findById(songId).ifPresent(song -> {
                song.setAlbum(null);
                songRepository.save(song);
            });
        }

        fileStorage.delete(album.getPicture());
        albumRepository.deleteById(id);
        return id;
    }

    public AlbumDocument addSong(String albumId, String songId) {
        AlbumDocument album = albumRepository.findById(albumId)
                .orElseThrow(() -> new ApiException(HttpStatus.NOT_FOUND, "Album not found"));
        SongDocument song = songRepository.findById(songId)
                .orElseThrow(() -> new ApiException(HttpStatus.NOT_FOUND, "Song not found"));

        if (!album.getSongs().contains(songId)) {
            album.getSongs().add(songId);
        }
        song.setAlbum(albumId);
        albumRepository.save(album);
        songRepository.save(song);
        return album;
    }

    public List<AlbumDocument> search(String query) {
        return albumRepository.searchByNameOrAuthor(query);
    }

    public long getCount() {
        return albumRepository.count();
    }
}
